### PR TITLE
Vracer2

### PR DESCRIPTION
**Pull request by SergioMartin86**
Merging vracer2 to master with all the changes we've accrued so far.

+ Improvements to all agents
+ New examples and study cases
+ new NN layer (Korali + Eigen3)

### CODE REVIEW

**Review by wadaniel — APPROVED** [review]

File: examples/learning/reinforcement/cartpole/.run_test.sh [hunk 1]
```diff
@@ -24,14 +24,15 @@ rm -rf __test-*; check_result
 
 for file in *.py
 do
- sed -e 's%Defining Termination Criteria%Defining Termination Criteria\ne["Solver"]["Termination Criteria"]["Max Generations"] = 40\n%g' \
+ sed -e 's%Defining Termination Criteria%Defining Termination Criteria\ne["Solver"]["Termination Criteria"]["Max Generations"] = 30\n%g' \
```
Comment (webepasc) [inline examples/learning/reinforcement/cartpole/.run_test.sh:27]: I guess this is still subject to ongoing work

File: examples/learning/reinforcement/cartpole/run-ddpg.py [hunk 2]
```diff
@@ -42,60 +42,50 @@
 
 ### Defining the configuration of replay memory
 
-e["Solver"]["Experience Replay"]["Start Size"] =   1000
-e["Solver"]["Experience Replay"]["Maximum Size"] = 100000
+e["Solver"]["Experience Replay"]["Start Size"] =   5000
```
Comment (webepasc) [inline examples/learning/reinforcement/cartpole/run-ddpg.py:45]: I'll just add one comment ---> Do all algos still solve the cartpole? What is the status?

File: source/auxiliar/math.hpp [hunk 3]
```diff
@@ -107,6 +148,126 @@ T dotProduct(const std::vector<T> &x, const std::vector<T> &y)
...
+* @param x denisty evaluation point
+* @param alpha Shape of Beta distribution
+* @param beta Shape of Beta distribution
+* @return The log density
+*/
+template <typename T>
+T betaLogDensity(const T &x, const T &alpha, const T &beta)
+{
+  T invBab = gsl_sf_lngamma(alpha) + gsl_sf_lngamma(beta) - gsl_sf_lngamma(alpha + beta);
+  return (alpha - 1.) * std::log(x) + (beta - 1.) * std::log(1. - x) * invBab;
```
Comment (webepasc) [inline source/auxiliar/math.hpp:177]: should't this be invBab*((alpha - 1.) * std::log(x) + (beta - 1.) * std::log(1. - x) ) ?? See https://en.wikipedia.org/wiki/Beta_distribution#Probability_density_function

File: source/modules/neuralNetwork/layer/dense/dense._cpp [hunk 4]
```diff
@@ -1 +1,913 @@
...
+
+#ifdef _KORALI_USE_EIGEN
+    if (_nn->_engine == "Korali")
+    {
+      // Performing Wx computation
+      Map<MatrixXf> matA(_weightValues, IC, OC);
+      Map<MatrixXf> matB(_prevLayer->_outputValues, IC, N);
+      Map<MatrixXf> matC(_inputValues, N, OC);
+
+      matC = matA.transpose() * matB;
```
Comment (webepasc) [inline source/modules/neuralNetwork/layer/dense/dense._cpp:348]: Could you elaborate why you can not use matA*matB with matA(_weightValues, OC, IC)?

File: source/modules/neuralNetwork/layer/dense/dense._cpp [hunk 5]
```diff
@@ -1 +1,913 @@
...
+
+#ifdef _KORALI_USE_EIGEN
+  if (_nn->_engine == "Korali")
+  {
+    // Backward propagating Wx+b operation
+    Map<MatrixXf> matA(_weightValues, IC, OC);
+    Map<MatrixXf> matB(_inputDiff, OC, N);
+    Map<MatrixXf> matC(_prevLayer->_outputDiff, N, IC);
+
+    matC = matA * matB;
```
Comment (webepasc) [inline source/modules/neuralNetwork/layer/dense/dense._cpp:600]: Ah because you want to have the backward nicely and not transpose the matrix here..?

**Review by webepasc — COMMENTED** [review]
Nice, maybe look at the comment for the Beta distribution. You can ignore the other comments I think 👍

**Pull request merged**